rts/Rendering/CommandDrawer.cpp

```diff
@@ -660,149 +665,83 @@ void CommandDrawer::DrawDefaultCommand(const Command& c, const CUnit* owner) con
 	lineDrawer.DrawLineAndIcon(dd->cmdIconID, unit->GetObjDrawErrorPos(owner->allyteam), dd->color);
 }
 
-void CommandDrawer::DrawQuedBuildingSquares(const CBuilderCAI* cai) const
+void CommandDrawer::ClearQueuedBuildingSquaresCache() { biCache.clear(); }
+
+void CommandDrawer::DrawQueuedBuildingSquares(const CBuilderCAI* cai, const SColor& color)
 {
 	const CCommandQueue& commandQue = cai->commandQue;
 	const auto& buildOptions = cai->buildOptions;
 
-	unsigned int  buildCommands = 0;
-	unsigned int uwaterCommands = 0;
+	auto& rb = RenderBuffer::GetTypedRenderBuffer<VA_TYPE_C>();
 
-	for (const Command& c: commandQue) {
-		if (buildOptions.find(c.GetID()) == buildOptions.end())
+	for (const Command& c : commandQue) {
+		#if 0
+		if (std::find(buildOptions.begin(), buildOptions.end(), c.GetID()) == buildOptions.end())
 			continue;
+		#endif
 
 		BuildInfo bi;
 
 		if (!bi.Parse(c))
 			continue;
 
-		bi.pos = CGameHelper::Pos2BuildPos(bi, false);
-
-		buildCommands += 1;
-		uwaterCommands += (bi.pos.y < 0.0f);
-	}
-
-	// worst case - 2 squares per building (when underwater) - 8 vertices * 3 floats
-	std::vector<GLfloat>   quadVerts(buildCommands * 12);
-	std::vector<GLfloat> uwquadVerts(buildCommands * 12); // underwater
-	// 4 vertical lines
-	std::vector<GLfloat> lineVerts(uwaterCommands * 24);
-	// colors for lines
-	std::vector<GLfloat> lineColors(uwaterCommands * 48);
-
-	unsigned int   quadcounter = 0;
-	unsigned int uwquadcounter = 0;
-	unsigned int   linecounter = 0;
+		//already drawn by other builder
+		BuildInfoHash biHasher; size_t biHash = biHasher(bi);
+		auto it = std::find_if(biCache.begin(), biCache.end(), [biHash, &bi](const auto& item) {
+			return (biHash == item.first) && (bi == item.second);
+		});
 
-	for (const Command& c: commandQue) {
-		if (buildOptions.find(c.GetID()) == buildOptions.end())
+		if (it != biCache.end())
 			continue;
 
-		BuildInfo bi;
-
-		if (!bi.Parse(c))
-			continue;
+		biCache.emplace_back(biHash, bi);
 
 		bi.pos = CGameHelper::Pos2BuildPos(bi, false);
-
 		const float xsize = bi.GetXSize() * (SQUARE_SIZE >> 1);
 		const float zsize = bi.GetZSize() * (SQUARE_SIZE >> 1);
+		const float radius = math::sqrt(xsize * xsize + zsize * zsize);
```

Comment:
Is this sufficient instead of the model radius? What if it's a tall tower that has a 1x1 footprint?